### PR TITLE
add LDAP specific API call to resolve an LDAP (login) name to an ownC…

### LINKED ISSUE #15084: Meta data missing the username when querying for an ldap user via the provisioning api
### Steps to reproduce

1. Install owncloud
2. configure ldap 
3. query for an ldap user using the provisioning api

### Expected behaviour

Expect all data to be returned

### Actual behaviour

The "Users" field is missing the username

```
jennifer@ubuntu-14-client:~/smash-box/lib/oc-tests$ curl http://<user>:<pw>@x.x.x.x/owncloud/ocs/v1.php/cloud/users
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data>
  <users>
   <element>aaliyah_oconnell_II</element>
   <element>abagail_kuhic</element>
   <element>abagail_kuhlman_III</element>
   <element>abagail_runolfsdottir</element>
   <element>at_admin</element>
   <element>user1</element>
   <element>user2</element>
   <element>user3</element>
  </users>
 </data>
</ocs>
jennifer@ubuntu-14-client:~/smash-box/lib/oc-tests$ curl http://<user>:<pw>@x.x.x.x/owncloud/ocs/v1.php/cloud/users?search=user1
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data>
  <users>
   <element>user1</element>
  </users>
 </data>
</ocs>
jennifer@ubuntu-14-client:~/smash-box/lib/oc-tests$ curl http://<user>:<pw>@x.x.x.x/owncloud/ocs/v1.php/cloud/users?search=abagail_kuhic
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data>
  <users/>               
 </data>
</ocs>
jennifer@ubuntu-14-client:~/smas-hbox/lib/oc-tests$


```

### Server configuration

Ubuntu 14.04
PHP 5.5.9
MySQL DB

ownCloud Enterprise Edition 8.1 pre alpha (daily) Build:2015-03-19T03:27:00+00:00 fe8002a7db0ccfec3a96def4f889c265b52ace53

### PR DESCRIPTION
**Pull request by blizzz**
…loud username that can be used with user related API calls

Fixes #15084 

This introduces the cloud/apps/user_ldap/resolveloginname/{loginname} route to the  Provisioning  API as suggested in https://github.com/owncloud/core/issues/15084#issuecomment-142245453.

Following output is created on those cases:

```
# unknown user
$ curl http://master:master@zara.owncloud.bzoc/master/ocs/v1.php/cloud/apps/user_ldap/resolveloginname/foobar
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>998</statuscode>
  <message>Could not resolve login name to user name</message>
 </meta>
 <data/>
</ocs>

# success
$ curl http://master:master@zara.owncloud.bzoc/master/ocs/v1.php/cloud/apps/user_ldap/resolveloginname/alice
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data>
  <element>dc35a74a-5de6-1932-93af-e3cc75610c15</element>
 </data>
</ocs>

# LDAP enabled, but no active configurations
$ curl http://master:master@zara.owncloud.bzoc/master/ocs/v1.php/cloud/apps/user_ldap/resolveloginname/alice
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>996</statuscode>
  <message>No active LDAP configuration</message>
 </meta>
 <data/>
</ocs>

# LDAP not enabled
$ curl http://master:master@zara.owncloud.bzoc/master/ocs/v1.php/cloud/apps/user_ldap/resolveloginname/alice
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>996</statuscode>
  <message>LDAP backend is not enabled</message>
 </meta>
 <data/>
</ocs>
```

Please test and review @tomneedham @rullzer @MorrisJobke

### CODE REVIEW

**Comment by karlitschek** [comment]
makes sense :+1:

File: apps/provisioning_api/lib/apps.php [hunk 1]
```diff
@@ -105,4 +105,44 @@ public function disable($parameters) {
...
+	 */
+	public function resolveLDAPLoginName($parameters) {
+		$loginName = $parameters['loginname'];
+
+		if(!$this->appManager->isEnabledForUser('user_ldap')) {
+			return new OC_OCS_Result(null, \OCP\API::RESPOND_SERVER_ERROR, 'LDAP backend is not enabled');
+		}
+
+		try {
+			$helper = new \OCA\user_ldap\lib\Helper();
```
Comment (DeepDiver1975) [inline apps/provisioning_api/lib/apps.php]: :-1: we should not reuse ldap classes in the provisioning app

Comment (DeepDiver1975) [inline apps/provisioning_api/lib/apps.php]: please move this code to the ldap app - no need to keep it in the provisioing app - THX

Comment (blizzz) [inline apps/provisioning_api/lib/apps.php]: @DeepDiver1975 now, is this more what you were looking for? Or do you prefer to keep it totally out of the Provisioning API? Having it there is what was agreed on, but I can put it anywhere of course.

File: apps/provisioning_api/lib/apps.php [hunk 2]
```diff
@@ -33,11 +33,17 @@ class Apps {
...
+	private $ldapBackend;
+
 	/**
 	 * @param \OCP\App\IAppManager $appManager
+	 * @param \OCA\user_ldap\User_Proxy|null $ldapBackend
 	 */
-	public function __construct(\OCP\App\IAppManager $appManager) {
-		$this->appManager = $appManager;
+	public function __construct(\OCP\App\IAppManager $appManager,
+								\OCA\user_ldap\User_Proxy $ldapBackend = null) {
```
Comment (nickvergessen) [inline apps/provisioning_api/lib/apps.php]: Inter app dependency? Just create your own OCS endpoint?

**Comment by nickvergessen** [comment]
I think it would be also okay to have this in general (placed in the user manager)
Because for the database backend it would also be good to have an endpoint to get the correct casing of a user id, ref https://github.com/owncloud/core/pull/19580

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 now, is this more what you were looking for? Or do you prefer to keep it totally out of the Provisioning API? Having it there is what was agreed on, but I can put it anywhere of course.

Having this OCS api in the scope of provisioning API is give by the URL of the http call.
The code itself to be located in ldap otherwise we totally violate our principals.

**Comment by DeepDiver1975** [comment]
Having a look at this and other places this all looks like workaround over workaround.

I'd vote for postponing this until 9.0 and see if there is anything we can do in a more general and centric approach

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque @MTRichards we might want to setup a story around user mgmt - there are some issue piling up which needs to be adresses.

**Comment by MTRichards** [comment]
Ok. Can you say more - What else is missing / needs work?

File: apps/provisioning_api/lib/apps.php (hunk 2 above)
Comment (icewind1991) [inline apps/provisioning_api/lib/apps.php]: This sounds like it will break when the ldap backend is not installed

**Comment by MorrisJobke** [comment]
> I'd vote for postponing this until 9.0 and see if there is anything we can do in a more general and centric approach

cc @karlitschek @cmonteroluque

**Comment by DeepDiver1975** [comment]
> Ok. Can you say more - What else is missing / needs work?

- uppercase/lowercase user name issues all over the place
- have ONE centric user accounts table to rule them all
- reliable calculation of users, data dirs regardless of auth backends being available/online or not
- reliable enumeration of users (e.g. run occ commands for all users - it's a mess today)

**Comment by MorrisJobke** [comment]
> reliable enumeration of users (e.g. run occ commands for all users - it's a mess today)

Additional: Having a properly sorted list of users. Currently the ldap users are sorted in the JS part, because LDAP doesn't support this kind of sorting.

**Comment by cmonteroluque** [comment]
@DeepDiver1975 and @MorrisJobke I am worried setting up this story will end up with it in 9.1 and that seems to far forward  (I don't see a good path for a faster timeline). At this point I'd prefer a good workaround AND a story for the future.

@karlitschek @MTRichards your thoughts?

**Comment by karlitschek** [comment]
Hmm. What is the severity here? Anyone waiting for this fix? @blizzz How big is the risk if we put this into 8.2?

**Comment by blizzz** [comment]
@karlitschek there's nothing agreed on that we can mere. I could move this part to the LDAP backend solely (then we have an API call and could not throw it away for awhile…), now it was suggested to put it on hold https://github.com/owncloud/core/pull/19600#issuecomment-146505795

**Comment by cmonteroluque** [comment]
ok, this has to be addressed in 9.0 @MTRichards @karlitschek

**Comment by MTRichards** [comment]
Ok.

**Comment by PVince81** [comment]
> ok, this has to be addressed in 9.0

The scope needs to be cleared out.
Are we talking about @blizzz's workaround of adding an API or about the non-trivial topics from @DeepDiver1975 here https://github.com/owncloud/core/pull/19600#issuecomment-146842644 ?

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/21282 to discuss the "central user lookup table" approach.

**Comment by PVince81** [comment]
- This adds new APIs and we're past feature freeze again
- It was raised originally by QA (Jennifer), so no one seems to be really waiting for this
- As per @DeepDiver1975, this is workaround over workaround and we'd better look into having the central users table as per https://github.com/owncloud/core/issues/21282 (9.1)

I vote for deferring this to 9.1.

**Comment by MorrisJobke** [comment]
:city_sunset: it!

**Comment by blizzz** [comment]
not needed

**Pull request closed without merging**